src/gg.regression.unity.bots/Runtime/Scripts/Validation/RGValidateBehaviour.cs

```diff
@@ -0,0 +1,240 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using StateRecorder.BotSegments.Models.SegmentValidations;
+using UnityEngine;
+using Object = UnityEngine.Object;
+
+namespace RegressionGames.Validation
+{
+    public class RGValidateBehaviour: MonoBehaviour
```

Comment:
We should model this differently. Instead of spawning monobehaviours and running into timing issues where endCriteria match before we've even spawned these, why not handle instantiation/evaluation as part of our botsegment evaluation loop like actions/endConditions are done ?

So instead of these being behaviours that self process on the Update loop, they would be implementations of an interface that a ValidationEvaluator behaviour driven by our botsegment update loop would process through on each update pass.

This would remove your ambiguous timing issues.

The one user code difference is that these wouldn't be MonoBehaviour objects anymore, but given that we aren't parenting these to anything useful anyway, I see that as an improvement. The APIs the user needs to find objects are still available to them MonoBehaviour or not because we'll be calling them from the main thread Update loop.

Reply:
I was going to counter with why I want it to be a monobehaviour, but after going through the functions and properties available in that class, I suppose I can't support my claim... I was thinking maybe they'd want access to things like mouse event handlers, etc... but with the opinionated way I want them to use this class, they shouldn't do that.

TLDR; I'll try out your approach since it will simplify some of the internals as well. Easy to change in the future if needed. I put what I was writing below so you can see my thought process that I abandoned while writing lol

----


Hmmm that is one option... yeah I was primarily concerned with the ability to use things that monobehaviours make available with ease.

But I guess to clarify, I feel like my timing issues weren't because of the update loop - actually, the suggested approach is somewhat equivalent but I am just using another gameobject's update loop to call my class's fake update loops, instead of having unity call it directly (since Unity goes through all update loops serially).

For "they would be implementations of an interface that a ValidationEvaluator class driven..." that also has been added, the validation segments can use the `ProcessValidations` function for that, and my intent here is that most of the types would - it just didn't for this one.

Reply:
This has now been updated to use the update loop of the bot segment playback controller 👍